### PR TITLE
Fixes #820: Registration fails for Categoricals created with .from_codes()

### LINKED ISSUE #820: Registration fails for Categoricals created with `.from_codes()`
The `.register()` method of `ak.Categorical` fails when the Categorical was created using `ak.Categorical.from_codes()`. This is because that constructor does not set the `.permutation` and `.segments` attributes, so when the `.register()` method tries to register them, it is attempting to register a `None`. The following code snippet reproduces the bug:

```python
>>> categories = ak.array(['a', 'b', 'c'])
>>> codes = ak.array([0, 1, 0, 2, 1])
>>> cat = ak.Categorical.from_codes(codes, categories)
>>> cat.register('cat')

AttributeError
```

I think we should have `.register`/`.attach` treat `.permutation` and `.segments` as optional attributes, i.e. ignore if not present.

### PR DESCRIPTION
**Pull request by pierce314159**
This PR (Fixes #820):
- Adds `Categorical.RequiredPieces` which contains `'categories'` and `'codes'`
- Adds an internal `_get_components`  method which returns all required and non-`None` components (limits duplicated logic)
- Modifies Categorical registration suite to handle when `permutations` or `segments` are `None`
- Adds a test of the Categorical registration suite on `Categorical` made using `.from_codes`

### CODE REVIEW

File: arkouda/categorical.py [hunk 1]
```diff
@@ -595,7 +596,8 @@ def register(self, user_defined_name:str) -> Categorical:
         Objects registered with the server are immune to deletion until
         they are unregistered.
         """
-        [getattr(self, p).register(f"{user_defined_name}.{p}") for p in Categorical.RegisterablePieces]
+        [getattr(self, p).register(f"{user_defined_name}.{p}") for p in Categorical.RegisterablePieces
+         if p in Categorical.RequiredPieces or getattr(self, p) is not None]
```
Comment (hokiegeek2) [inline arkouda/categorical.py:600]: I recommend putting this logic within a python closure method and adding some pydoc comments

File: arkouda/categorical.py [hunk 2]
```diff
@@ -647,12 +649,16 @@ def is_registered(self) -> np.bool_:
...
-        parts_registered:List[np.bool_] = [getattr(self, p).is_registered() for p in Categorical.RegisterablePieces]
+        parts_registered: List[np.bool_] = [p.is_registered() for p in Categorical._get_components(self)]
         if np.any(parts_registered) and not np.all(parts_registered):  # test for error
             raise RegistrationError(f"Not all registerable components of Categorical {self.name} are registered.")
 
         return np.bool_(np.any(parts_registered))
 
+    def _get_components(self) -> List:
+        return [getattr(self, p) for p in Categorical.RegisterablePieces if p in Categorical.RequiredPieces
+                or getattr(self, p) is not None]
```
Comment (hokiegeek2) [inline arkouda/categorical.py:660]: ditto for this. The code is fine and I understand it, but a helper method with a defined interface coupled with pydoc comments would be helpful for a new Arkouda developer

File: arkouda/categorical.py [hunk 3]
```diff
@@ -666,7 +672,8 @@ def _list_component_names(self) -> List[str]:
         List[str]
             List of all component names
         """
-        return list(itertools.chain.from_iterable([getattr(self, p)._list_component_names() for p in Categorical.RegisterablePieces]))
+        return list(itertools.chain.from_iterable(
```
Comment (hokiegeek2) [inline arkouda/categorical.py:675]: use of itertools is nice -> well done

**Review by hokiegeek2 — APPROVED** [review]
Looks really good to me, just some minor suggestions -> good job Pierce!

**Review by mhmerrill — APPROVED** [review]
changes look ok to me

**Comment by mhmerrill** [comment]
@pierce314159 @hokiegeek2  I am going to merge this, any further changes can be put in a future PR ;-)

**Comment by hokiegeek2** [comment]
👍

**Pull request merged**